src/Control/Timeout.hs

```diff
@@ -0,0 +1,103 @@
+
+module Control.Timeout (
+  Cancel,
+  shouldCancelRightNow,
+  checkForCancel,
+  Duration (..),
+  durationToMicro,
+  timeout',
+  timeoutIO,
+) where
+
+import Control.Carrier.Threaded (fork, kill)
+import Control.Concurrent (
+  isEmptyMVar,
+  newEmptyMVar,
+  putMVar,
+  takeMVar,
+  threadDelay,
+ )
+import Control.Concurrent.Async qualified as Async
+import Control.Effect.Diagnostics (
+  Diagnostics,
+  ToDiagnostic,
+  fatal,
+ )
+import Control.Effect.Exception (finally)
+import Control.Effect.Lift (Has, Lift, sendIO)
+import Control.Monad (when)
+import Control.Timeout.Internal (
+  Cancel (..),
+  Duration (..),
+  durationToMicro,
+ )
+import Data.Functor (($>))
+
+-- Whether a cancellable should cancel right now, or poll again.
+-- Use checkForCancel instead if you want to fail.
+shouldCancelRightNow :: Cancel -> IO Bool
+shouldCancelRightNow (Cancel mvar) = not <$> isEmptyMVar mvar
+
+-- | @checkForCancel err cancel@ will return @fatal err@
+-- if @shouldCancelRightnow cancel@ returns true.
+checkForCancel ::
+  ( Has Diagnostics sig m
+  , Has (Lift IO) sig m
+  , ToDiagnostic err
+  ) =>
+  err ->
+  Cancel ->
+  m ()
+checkForCancel err cancel = do
+  should <- sendIO $ shouldCancelRightNow cancel
+  when should $ fatal err
+
+-- | @timeout' duration act@ will allow internal cancellation functions to be
+-- used, by passing a cancellation token ('Cancel') to act.
+-- A negative duration is the same as no time, and will set the cancel flag
+-- almost immediately.
+--
+-- The intended usage is:
+-- @
+-- timeout' duration $ \cancelToken -> do
+--   longRunningComputation cancelFlag
+-- @
+--
+-- Which will run @longRunningComputation@ until it returns, and set the cancel
+-- flag after the duration expires.  For a cancellation to occur,
+-- @longRunningComputation@ must return a value early by checking the cancel
+-- flag with 'checkForCancel'.  If @longRunningComputation@ does not check the
+-- cancel flag, or doesn't return early when the flag is set, no timeout will
+-- occur.
+timeout' :: (Has (Lift IO) sig m) => Duration -> (Cancel -> m a) -> m a
+timeout' duration act = do
+  -- We start with an empty MVar to signal that we're not ready to cancel.
+  cancel <- sendIO newEmptyMVar
+  start <- sendIO newEmptyMVar
+  handle <- sendIO . fork $ do
+    -- signal that the fork is about to begin it's sleep
+    putMVar start ()
+    threadDelay $ durationToMicro duration
+    -- We fill the MVar here, which is the cancel signal.
+    -- SAFETY: putMVar can block infinitely if we do it twice, since we never
+    --   empty the MVar. Since we never expose the underlying MVar to the
+    --   consumer, we only have to validate that this putMVar call is safe,
+    --   and that all other usage in this module is non-blocking.
+    putMVar cancel ()
+  -- Once 'start' is taken, we know that the fork has begun the timer
+  -- We have to actually wait for this value, which is why we use $!.
+  sendIO $! takeMVar start
```

Comment:
can we eliminate the `start` var entirely / just assume the other thread is doing the thing?

Reply:
unreliable on what timescale? if the thread is sleeping for anything more than a few ms, I don't expect this wait-for-start to substantially impact the timing

Reply:
I was testing at 1ms scales, that's a fair point, our timing is never nearly that fine-grained.  I'll retest it at 200ms-2s scales.  If it's not that substantial, we can just remove this.

Reply:
Actually, even that is too small for what we normally do.  I can do manual testing at a 10s scale, and leave unit tests for about 200ms-1s.  At that point, there's almost no way this is actually needed.